keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1000 @@
+output is associated with the test case.
+
+## Motivation
+
+### Goals
+
+- Remove `k8s.io/klog` imports and usage from all packages.
+- Grant the caller of a function control over logging inside that function.
+- Provide documentation and helper code for setting up logging in unit tests.
+- Change as few exported APIs as possible.
+
+### Non-Goals
+
+- Remove the klog text output format.
+
+## Proposal
+
+Log calls that were already converted to structured logging need to be updated
+to use a logger that gets passed into the function in one of two ways:
+- as explicit parameter
+- attached to a `context.Context`
+
+When a function already accepts a context parameter, then that will be used
+instead of adding a separate parameter. This covers most of client-go and
+avoids another big breaking change for the community.
+
+When a function does not accept a context parameter but needs a context for
+calling some other functions, then a context parameter will be added. As a
+positive side effect, such a change can then also remove several `context.TODO`
+calls (currently over 6000 under `pkg`, `cmd`, `staging` and `test`). An
+explicit logger parameter is suitable for functions which don’t need a context
+and never will.
+
+The rationale for not using both context and an explicit logger parameter is
+the risk that the caller passes an updated logger which it didn't add to the
+context. When the context is then passed down to other functions instead of the
+logger, logging in those functions will not produce the expected result. This
+could be avoided by carefully reviewing code which modifies loggers, but
+designing an API so that such a mistake cannot happen seems safer. The logcheck
+linter will check for this. `nolint:logcheck` comments can be used for
+those functions where passing both is preferred despite the ambiguity.
+
+`klog.KObj` and similar helper functions get moved to `k8s.io/klogr`. That package
+will also provide aliases for functions and types from `go-logr/logr`. That way,
+a single import statement will be enough in most files. It will also implement
+loggers for klog text output and unit testing.
+
+When a function retrieves a logger from a context parameter and no logger has
+been set there, it will fall back transparently to logging through a global
+logger instance tracked in `k8s.io/klogr`, i.e. log calls never need to check
+whether they have a valid logger.
+
+### User Stories
+
+#### Story 1
+
+kube-scheduler developer Joan [wants to know which
+pod](https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-675074671)
+and which operation and scheduler plugin log messages are associated with.
+
+When kube-scheduler starts processing a pod, it creates a new logger with
+`logger.WithValue("pod", klogr.KObj(pod))` and passes that around. While
+iterating over plugins in certain operations, another logger gets created with
+`logger.WithName(<operation>).WithName(<plugin name>)` and then is used when
```

Comment:
What guidance we want to make for devs picking `WithName` instead of `WithValue`. In this example passing plugin name will make it harder to group by logs by plugin when compared to using `WithValue("plugin", <plugin name>)`.

Reply:
I don't have any strong preference. We can discuss that when writing the documentation.

Reply:
SG